File: c/Sample2.c
#include <assert.h>
#include <stdbool.h>
struct {int coverage[10000];} TestCoverage;
int XBF_WRITE = 1;
int XBF_ASYNC = 2;
int XBF_READ = 4;
int _XBF_DELWRI_Q = 8;
int XBF_WRITE_FAIL = 16;
int XBF_DONE = 32;

int SHUTDOWN_META_IO_ERROR = 1;

bool assert_ok = false;

void ASSERT(bool b) {
  assert_ok = b;
}

struct b_target {
  int bt_mount;
};

struct xfs_buf {
  int b_flags;
  struct b_target *b_target;
  int is_locked;
  int wait_error;
};

int line_no;
int xfs_buf_islocked(struct xfs_buf *bp) {
  assert(line_no == 0);
  line_no++;
  return bp->is_locked;
}
int xfs_buf_submit_wait(struct xfs_buf *bp) {
  assert(line_no == 1);
  line_no++;
  return bp->wait_error;
}
void xfs_force_shutdown(int* bt_mount, int error) {
  assert(line_no == 2);
  line_no++;
  *bt_mount = error;
}

#START OF SAMPLE
int xfs_bwrite(struct xfs_buf    *bp)
{
  int      error;

  ASSERT(xfs_buf_islocked(bp));

  bp->b_flags |= XBF_WRITE;
  bp->b_flags &= ~(XBF_ASYNC | XBF_READ | _XBF_DELWRI_Q |
      XBF_WRITE_FAIL | XBF_DONE);

  error = xfs_buf_submit_wait(bp);
  if (error) {
    xfs_force_shutdown(&bp->b_target->bt_mount,
        SHUTDOWN_META_IO_ERROR);
  }
  return error;
}
#END OF SAMPLE

void test_good() {
  struct xfs_buf bp;
  struct b_target t;
  bp.b_target = &t;
  bp.b_target->bt_mount = 0;
  bp.b_flags = XBF_ASYNC | XBF_READ | _XBF_DELWRI_Q | XBF_WRITE_FAIL | XBF_DONE;
  bp.is_locked = true;
  bp.wait_error = false;
  line_no = 0;
  assert(!xfs_bwrite(&bp));
  assert(bp.b_flags = XBF_WRITE);
  assert(assert_ok);
  assert(bp.b_target->bt_mount == 0);
  assert(line_no == 2);
}
void test_not_locked() {
  struct xfs_buf bp;
  struct b_target t;
  bp.b_target = &t;
  bp.is_locked = false;
  line_no = 0;
  assert(!xfs_bwrite(&bp));
  assert(!assert_ok);
  assert(line_no == 2);
}
void test_wait_error() {
  struct xfs_buf bp;
  struct b_target t;
  bp.b_target = &t;
  bp.b_target->bt_mount = 0;
  bp.is_locked = true;
  bp.wait_error = true;
  line_no = 0;
  assert(xfs_bwrite(&bp));
  assert(bp.b_target->bt_mount == SHUTDOWN_META_IO_ERROR);
  assert(line_no == 3);
}

int main() {
  test_good();
  test_not_locked();
  test_wait_error();
}

File: lua/Sample2.lua
TestCoverage = {coverage={}}

my_error = 6426

old_assert = assert

function asrt(b)
    old_assert(b, "")
end

expected_truth = false

assert_called = false

function assert(b, s)
    asrt(b == expected_truth)
    asrt(s == 'Expected n >= 0!')
    assert_called = true

    if not b then
        error(my_error)
    end
end

#START OF SAMPLE
local function fib_iter(n)
    assert(n > 0, 'Expected n >= 0!')
    return coroutine.wrap(function()
        local a, b, k = 0, 1, 0
        while k < n do
            a, b = b, a+b
            k = k + 1
            coroutine.yield(k, a)
        end
    end)
end
#END OF SAMPLE

expected_truth = false
status, ret = pcall(fib_iter, -1)
asrt(status == false)
asrt(ret == my_error)
asrt(assert_called == true)


expected_truth = true
assert_called = false
f = fib_iter(6)
asrt(assert_called == true)
x, y = f()
asrt(x == 1 and y == 1)
x, y = f()
asrt(x == 2 and y == 1)
x, y = f()
asrt(x == 3 and y == 2)
x, y = f()
asrt(x == 4 and y == 3)
x, y = f()
asrt(x == 5 and y == 5)
x, y = f()
asrt(x == 6 and y == 8)

asrt(f() == nil)